### DISCUSSION TITLE
Shelly 3EM PRO

### DISCUSSION THREAD

**Discussion by maxi673** [community/community#134809] [Programming Help]
Good day,
Is anyone here kind enough to help me create a Shelly script for the 3EM PRO? I would be very happy.
I would need the following function:
I have a Shelly 3 EM PRO and use it to measure the total consumption of my house. If the total power of my house falls below -2500W, the output of the Switch Add-On should be switched on. (PV on the roof and the Add-On switches on a heating element). It should then be possible to set a minimum runtime (minimum time that the heating element remains switched on regardless of the power) and a switch-on delay. The heating element has an output of 1500W. This means that as soon as it is switched on, I only have -1000W going into the mains.  If the mains-related power rises above -500W again and the minimum runtime has expired, the immersion heater should be switched off again and given a restart delay for 5 minutes.
I would be very grateful for any help.

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @maxi673 ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by geniuszly** [comment]
```javascript
// Define configuration parameters
let min_runtime = 600000; // Minimum runtime in milliseconds (e.g., 10 minutes)
let switch_on_delay = 30000; // Switch-on delay in milliseconds (e.g., 30 seconds)
let restart_delay = 300000; // Restart delay in milliseconds (e.g., 5 minutes)

// State variables
let is_switch_on = false;
let switch_on_time = 0;
let last_off_time = 0;

// Function to monitor and control the Switch Add-On
Timer.set(10000, true, function() {
  // Read the total power consumption from Shelly 3EM
  let total_power = Shelly.getComponentStatus("emeter", "0").power + 
                    Shelly.getComponentStatus("emeter", "1").power + 
                    Shelly.getComponentStatus("emeter", "2").power;

  // Check if the power falls below -2500W and switch is off
  if (total_power < -2500 && !is_switch_on && (Sys.now() * 1000 - last_off_time) > restart_delay) {
    Timer.set(switch_on_delay, false, function() {
      Shelly.call("Switch.Set", { id: 0, on: true });
      is_switch_on = true;
      switch_on_time = Sys.now() * 1000;
    });
  }

  // Check if the power rises above -500W after minimum runtime has expired
  if (is_switch_on && (total_power > -500) && (Sys.now() * 1000 - switch_on_time) > min_runtime) {
    Shelly.call("Switch.Set", { id: 0, on: false });
    is_switch_on = false;
    last_off_time = Sys.now() * 1000;
  }
});
```
- `min_runtime`: Minimum time (in milliseconds) the heating element should stay on regardless of power conditions. Set to 10 minutes (`600000 ms`).
- `switch_on_delay`: The delay before the heating element is switched on once the condition is met. Set to 30 seconds (`30000 ms`).
- `restart_delay`: The time delay before the system allows the heating element to be turned on again after being switched off. Set to 5 minutes (`300000 ms`).
- `total_power`: The script reads the total power consumption by summing the power readings from all three phases (`emeter 0`, `emeter 1`, `emeter 2`).
- `Switch-On Condition`: The heating element is switched on when the total power drops below `-2500W`, and the system is not within the restart delay period.
- `Switch-Off Condition`: The heating element is turned off when the total power exceeds `-500W`, and the minimum runtime has elapsed.

This should provide the functionality you need to manage your heating element based on the power readings from the Shelly 3EM PRO.
